src/data/dfinsupp/well_founded.lean

```diff
@@ -0,0 +1,190 @@
+/-
+Copyright (c) 2022 Junyan Xu. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Junyan Xu
+-/
+import data.dfinsupp.lex
+import order.game_add
+import order.antisymmetrization
+import set_theory.ordinal.basic
+
+/-!
+# Well-foundedness of the lexicographic and product orders on `dfinsupp` and `pi`
+
+The proof is modelled on the proof of `well_founded.cut_expand`.
+-/
+
+variables {ι : Type*} {α : ι → Type*}
+
+namespace dfinsupp
+
+variables [hz : Π i, has_zero (α i)] (r : ι → ι → Prop) (s : Π i, α i → α i → Prop)
+include hz
+
+open relation prod
+
+lemma lex_fibration : fibration
+  (inv_image (game_add (dfinsupp.lex r s) (dfinsupp.lex r s)) snd)
+  (dfinsupp.lex r s)
+  (λ x, merge x.1 x.2.1 x.2.2) :=
+begin
+  rintro ⟨p, x₁, x₂⟩ x ⟨i, hr, hs⟩,
+  simp_rw [merge_apply] at hs hr,
+  split_ifs at hs, classical,
+  work_on_goal 1
+  { refine ⟨⟨λ j, r j i → p j, merge (λ j, r j i) x₁ x, x₂⟩, game_add.fst ⟨i, _⟩, _⟩ },
+  work_on_goal 3
+  { refine ⟨⟨λ j, r j i ∧ p j, x₁, merge (λ j, r j i) x₂ x⟩, game_add.snd ⟨i, _⟩, _⟩ },
+  swap 3, iterate 2
+  { simp_rw merge_apply,
+    refine ⟨λ j h, if_pos h, _⟩,
+    convert hs,
+    refine ite_eq_right_iff.2 (λ h', (hr i h').symm ▸ _),
+    rw if_neg h <|> rw if_pos h },
+  all_goals { ext j, simp_rw merge_apply, split_ifs with h₁ h₂ },
+  { rw [hr j h₂, if_pos (h₁ h₂)] },
+  { refl },
+  { rw not_imp at h₁, rw [hr j h₁.1, if_neg h₁.2] },
+  { rw [hr j h₁.1, if_pos h₁.2] },
+  { rw [hr j h₂, if_neg (λ h', h₁ ⟨h₂, h'⟩)] },
+  { refl },
+end
+
+variables {r s}
+
+lemma lex.acc_of_single_erase [decidable_eq ι] {x : Π₀ i, α i} (i : ι)
+  (hs : acc (dfinsupp.lex r s) (single i (x i)))
+  (hu : acc (dfinsupp.lex r s) (x.erase i)) : acc (dfinsupp.lex r s) x :=
+begin
+  classical, rw ← merge_single_erase x i,
+  convert acc.of_fibration _ (lex_fibration r s) _,
+  work_on_goal 4 { refine ⟨_, _, _⟩ }, iterate 3 { refl },
+  exact inv_image.accessible snd (hs.prod_game_add hu),
+end
+
+variable (hbot : ∀ ⦃i a⦄, ¬ s i a 0)
+include hbot
+
+lemma lex.acc_zero : acc (dfinsupp.lex r s) 0 := acc.intro 0 $ λ x ⟨_, _, h⟩, (hbot h).elim
+
+lemma lex.acc_of_single [decidable_eq ι] [Π i (x : α i), decidable (x ≠ 0)] (x : Π₀ i, α i) :
+  (∀ i ∈ x.support, acc (dfinsupp.lex r s) $ single i (x i)) → acc (dfinsupp.lex r s) x :=
+begin
+  generalize ht : x.support = t, revert x, classical,
+  induction t using finset.induction with b t hb ih,
+  { intros x ht, rw support_eq_empty.1 ht, exact λ _, lex.acc_zero hbot },
+  refine λ x ht h, lex.acc_of_single_erase b (h b $ t.mem_insert_self b) _,
+  refine ih _ (by rw [support_erase, ht, finset.erase_insert hb]) (λ a ha, _),
+  rw [erase_ne (ha.ne_of_not_mem hb)],
+  exact h a (finset.mem_insert_of_mem ha),
+end
+
+variable (hs : ∀ i, well_founded (s i))
+include hs
+
+lemma lex.acc_single [decidable_eq ι] {i : ι} (a : α i)
+  (hi : acc (rᶜ ⊓ (≠)) i) : acc (dfinsupp.lex r s) (single i a) :=
```

Comment:
The use of the complementary relation seems to be a bit strange at first. Can you add a docstring that explains it?

Reply:
Explained in the module docstring of *dfinsupp/well_founded*.